File: blasd/copy.c

// Copyright (c) Harri Rautila, 2013

// This file is part of github.com/hrautila/gomas library. It is free software,
// distributed under the terms of GNU Lesser General Public License Version 3, or
// any later version. See the COPYING tile included in this archive.

#include "dtype.h"
#include "interfaces.h"
#include "matcpy.h"

void __vec_copy(mvec_t *X,  const mvec_t *Y, int N)
{
  register int i, kx, ky;
  register double f0, f1, f2, f3;

  // gcc compiles loop body to use different target XMM registers
  for (i = 0; i < N-3; i += 4) {
    f0 = Y->md[(i+0)*Y->inc];
    f1 = Y->md[(i+1)*Y->inc];
    f2 = Y->md[(i+2)*Y->inc];
    f3 = Y->md[(i+3)*Y->inc];
    X->md[(i+0)*X->inc] = f0;
    X->md[(i+1)*X->inc] = f1;
    X->md[(i+2)*X->inc] = f2;
    X->md[(i+3)*X->inc] = f3;
  }    
  if (i == N)
    return;

  // calculate indexes only once
  kx = i*X->inc;
  ky = i*Y->inc;
  switch (N-i) {
  case 3:
    X->md[kx] = Y->md[ky];
    kx++; ky++;
  case 2:
    X->md[kx] = Y->md[ky];
    kx++; ky++;
  case 1:
    X->md[kx] = Y->md[ky];
  }
}

void __blk_copy(mdata_t *A,  const mdata_t *B, int M, int N)
{
  copy_plain_mcpy1(A->md, A->step, B->md, B->step, M, N);
}

// M is rows in source and cols in destination, N is cols in source
void __blk_transpose(mdata_t *A,  const mdata_t *B, int M, int N)
{
  copy_trans4x1(A->md, A->step, B->md, B->step, M, N);
}

// Local Variables:
// indent-tabs-mode: nil
// End:

File: blasd/copy.go

// Copyright (c) Harri Rautila, 2013

// This file is part of github.com/hrautila/gomas package. It is free software,
// distributed under the terms of GNU Lesser General Public License Version 3, or
// any later version. See the COPYING tile included in this archive.

package blasd

// #cgo CFLAGS: -O3 -march=native -fomit-frame-pointer -ffast-math -Iinc -I.
// #cgo LDFLAGS: -lm
// #include "inc/interfaces.h"
import  "C"
import "unsafe"
import "github.com/hrautila/cmat"
import "github.com/hrautila/gomas"

func vcopy(X, Y *cmat.FloatMatrix, N int) {
    var x, y C.mvec_t

    if N == 0 {
        return
    }
    xr, _ := X.Size()
    x.md = (*C.double)(unsafe.Pointer(&X.Data()[0]))
    x.inc = C.int(1)
    if xr == 1 {
        x.inc = C.int(X.Stride())
    }
    yr, _ := Y.Size()
    y.md = (*C.double)(unsafe.Pointer(&Y.Data()[0]))
    y.inc = C.int(1)
    if yr == 1 {
        y.inc = C.int(Y.Stride())
    }
    C.__d_vec_copy(
        (*C.mvec_t)(unsafe.Pointer(&x)),
        (*C.mvec_t)(unsafe.Pointer(&y)),  C.int(N))
    return
}

func mcopy(A, B *cmat.FloatMatrix, M, N int) {
    var a, b C.mdata_t

    if M == 0 || N == 0 {
        return
    }
    a.md = (*C.double)(unsafe.Pointer(&A.Data()[0]))
    a.step = C.int(A.Stride())
    b.md = (*C.double)(unsafe.Pointer(&B.Data()[0]))
    b.step = C.int(B.Stride())
    C.__d_blk_copy(
        (*C.mdata_t)(unsafe.Pointer(&a)),
        (*C.mdata_t)(unsafe.Pointer(&b)),  C.int(M), C.int(N))
    return
}

func mtranspose(A, B *cmat.FloatMatrix, M, N int) {
    var a, b C.mdata_t
    if M == 0 || N == 0 {
        return
    }

    a.md = (*C.double)(unsafe.Pointer(&A.Data()[0]))
    a.step = C.int(A.Stride())
    b.md = (*C.double)(unsafe.Pointer(&B.Data()[0]))
    b.step = C.int(B.Stride())
    C.__d_blk_transpose(
        (*C.mdata_t)(unsafe.Pointer(&a)),
        (*C.mdata_t)(unsafe.Pointer(&b)),  C.int(M), C.int(N))
    return
}


func Copy(A, B *cmat.FloatMatrix, confs ...*gomas.Config) *gomas.Error {
    ar, ac := A.Size()
    br, bc := B.Size()
    avec := ar == 1 || ac == 1
    bvec := br == 1 || bc == 1
    if avec && bvec {
        if A.Len() <= B.Len() {
            return gomas.NewError(gomas.ESIZE, "Copy")
        }
        vcopy(A, B, A.Len())
        return nil
    }
    if ar != br || ac != bc {
        return  gomas.NewError(gomas.ESIZE, "Copy")
    }
    mcopy(A, B, ar, ac)
    return nil
}


func Transpose(A, B *cmat.FloatMatrix, confs ...*gomas.Config) *gomas.Error {
    ar, ac := A.Size()
    br, bc := B.Size()
    if ar != bc || ac != br {
        return gomas.NewError(gomas.ESIZE, "Transpose")
    }
    mtranspose(A, B, br, bc)
    return nil
}


// Local Variables:
// tab-width: 4
// indent-tabs-mode: nil
// End:
